tests/test_data_types.py

```diff
@@ -0,0 +1,280 @@
+
+
+class TestBoundedFloat:
+    """
+    Test BoundedFloat-specific functionality.
+    """
+
+    @pytest.mark.parametrize(
+        'dtype,should_pass',
+        [
+            # Valid dtypes
+            (float, True),
+            (np.float32, True),
+            (np.float64, True),
+            # Invalid dtypes
+            (int, False),
+            (np.int32, False),
+            (np.int64, False),
+            (str, False),
+            (bool, False),
+            (complex, False),
+        ],
+    )
+    def test_dtype_validation(self, dtype, should_pass):
+        """Test that only valid float dtypes are accepted."""
+        if should_pass:
+            BoundedFloat(dtype=dtype)  # Should not raise
+        else:
+            with pytest.raises(ValueError, match='Invalid dtype'):
+                BoundedFloat(dtype=dtype)
+
+    @pytest.mark.parametrize(
+        'test_value,shape,should_pass',
+        [
+            # Scalar NaN values
+            (float('nan'), None, True),
+            # Arrays with valid values and NaN
+            ([0.5, float('nan'), 0.8], ['*'], True),
+            ([0.0, 1.0, float('nan')], ['*'], True),
+            ([float('nan')], ['*'], True),
+            # Arrays with invalid non-NaN values (should fail)
+            ([0.5, 1.5, float('nan')], ['*'], False),
+            ([-0.1, float('nan'), 0.5], ['*'], False),
+        ],
+    )
+    def test_nan_handling(self, test_value, shape, should_pass):
+        """Test that NaN values are allowed and filtered from validation."""
+        bounded_float = setup_datatype_for_testing(
+            BoundedFloat().min(0).max(1), shape=shape
+        )
+
+        if should_pass:
+            result = bounded_float.normalize(test_value)
+            if isinstance(test_value, list):
+                # Check that result has same structure and NaN positions
+                assert len(result) == len(test_value)
+                for i, (orig, res) in enumerate(zip(test_value, result)):
+                    if np.isnan(orig):
+                        assert np.isnan(res), f'NaN not preserved at position {i}'
+                    else:
+                        assert res == orig, f'Value mismatch at position {i}'
+            else:
+                assert np.isnan(result), 'Scalar NaN not preserved'
+        else:
+            with pytest.raises(
+                ValueError, match=r'All non-NaN values must be in \[0, 1\]'
+            ):
+                bounded_float.normalize(test_value)
+
+    def test_float_specific_validation(self):
+        """Test float-specific validation behavior."""
+        bounded_float = setup_datatype_for_testing(
+            BoundedFloat().min(0).max(1), shape=None
+        )
+
+        test_values = [0.5, float('nan'), 0.8]
+        filtered = bounded_float._filter_values_for_validation(test_values)
+        assert filtered == [0.5, 0.8]  # NaN should be filtered out
+
+
+class TestBoundsStringRepresentation:
```

Comment:
Try add a test that `m_from_dict(section.m_to_dict())` to cover both deserialization and serialization.

Reply:
I've added them under `TestBoundNumber.test_serialization_and_reconstruction` and `TestNOMADIntegration`